app/controllers/boxes_controller.rb

```diff
@@ -51,21 +51,16 @@ def new
 
   def create
     return unless authorize_resource(@navigation_context.institution, CREATE_INSTITUTION_BOX)
-    @box_form = BoxForm.build(@navigation_context, box_params)
 
+    @box_form = BoxForm.build(@navigation_context, box_params)
     @box_form.batches = check_access(load_batches, READ_BATCH)
-    @box_form.samples = check_access(load_samples, READ_SAMPLE)
 
-    if @box_form.valid?
+    if @box_form.save
       @box_form.build_samples
-
-      if @box_form.save
-        redirect_to boxes_path, notice: "Box was successfully created."
-        return
-      end
+      redirect_to boxes_path, notice: "Box was successfully created."
+    else
```

Comment:
> we save the box which will create a transaction and create the Box and its Samples in a transaction, so we can never have a Box without Samples (it caused bugs previously) or a Box with only some of the expected Samples (worst case).

I was thinking, we could move validation Box instead of BoxForm. What do you think?

```
class Box < ApplicationRecord
  has_many :samples
  validate :ensure_samples_validity

  private

  def ensure_samples_validity
    errors.add(:base, "Box must have at least one sample") if samples.empty?
    errors.add(:base, "You can't select a QC sample") if samples.any?(&:is_quality_control?)
  end
end


```

Reply:
Hum, having data integrity right into Box is tempting...

But validations are different depending on the selected option to create the box (there are too many ways to create a box).

- The "can't select QC sample" would be fine, since it can only happen when we manually select a sample.
- The "at least one sample message", on the other hand, would be wrong for the `add_batches` option (we only care about the selected batches), but also for the `add_csv` option (the wording is slightly different).

To discriminate, we'd need to push `@option` to Box, which would defeat the point of BoxForm (keep the creation logic from the HTML form distinct from the Box object).

Reply:
make sense, thanks for the clarification